c_binding/libsg.c

```diff
@@ -0,0 +1,488 @@
+    return rc;
+}
+
+static struct _sg_t10_vpd83_dp *_sg_t10_vpd83_dp_new(void)
+{
+    struct _sg_t10_vpd83_dp *dp = NULL;
+
+    dp = (struct _sg_t10_vpd83_dp *) malloc(sizeof(struct _sg_t10_vpd83_dp));
+
+    if (dp != NULL) {
+        memset(dp, 0, sizeof(struct _sg_t10_vpd83_dp));
+    }
+    return dp;
+}
+
+/*
+ * Assuming input pointer is not NULL, caller should do that.
+ */
+void _sg_t10_vpd83_dp_array_free(struct _sg_t10_vpd83_dp **dps,
+                                 uint16_t dp_count)
+{
+    uint16_t i = 0;
+    for (; i < dp_count; ++i) {
+        free(dps[i]);
+    }
+    free(dps);
+}
+
+/*
+ * Assuming err_msg and data are not NULL, caller should make sure that.
+ */
+int _sg_io_recv_diag(char *err_msg, int fd, uint8_t page_code, uint8_t *data)
+{
+    int rc = LSM_ERR_OK;
+    uint8_t cdb[_T10_SPC_RECV_DIAG_CMD_LEN];
+    int ioctl_errno = 0;
+
+    /* SPC-5 rev7, Table 219 - RECEIVE DIAGNOSTIC RESULTS command */
+    cdb[0] = RECEIVE_DIAGNOSTIC;                /* OPERATION CODE */
+    cdb[1] = 1;                                 /* PCV */
+    /* ^ We have no use case need PCV = 0 yet.
+     *   When PCV == 0, it means page code is invalid, just retrieve the result
+     *   of recent SEND DIAGNOSTIC.
+     */
+    cdb[2] = page_code & UINT8_MAX;             /* PAGE CODE */
+    cdb[3] = (_SG_T10_SPC_RECV_DIAG_MAX_LEN >> 8 ) & UINT8_MAX;
+                                                /* ALLOCATION LENGTH, MSB */
+    cdb[4] = _SG_T10_SPC_RECV_DIAG_MAX_LEN & UINT8_MAX;
+                                                /* ALLOCATION LENGTH, LSB */
+    cdb[5] = 0;                                 /* ^ CONTROL */
+    /* ^ We have no use case need to handle auto contingent allegiance(ACA)
+     *   yet.
+     */
+
+    ioctl_errno = _sg_io(fd, cdb, _T10_SPC_RECV_DIAG_CMD_LEN, data,
+                         _SG_T10_SPC_RECV_DIAG_MAX_LEN, _SG_IO_RECV_DATA);
+    if (ioctl_errno != 0) {
+        rc = LSM_ERR_LIB_BUG;
+        /* TODO(Gris Ge): Check 'Supported Diagnostic Pages diagnostic page' */
+
+        _lsm_err_msg_set(err_msg, "Got error from SGIO RECEIVE_DIAGNOSTIC "
+                         "for page code 0x%02x: %d(%s)", page_code,
+                         ioctl_errno, strerror(ioctl_errno));
+        goto out;
+    }
+
+ out:
+
+    return rc;
+}
+
+int _sg_io_send_diag(char *err_msg, int fd, uint8_t *data, uint16_t data_len)
+{
+    int rc = LSM_ERR_OK;
+    uint8_t cdb[_T10_SPC_SEND_DIAG_CMD_LEN];
+    int ioctl_errno = 0;
+
+    /* SPC-5 rev7, Table 219 - RECEIVE DIAGNOSTIC RESULTS command */
+    cdb[0] = SEND_DIAGNOSTIC;                   /* OPERATION CODE */
+    cdb[1] = 0x10;                              /* SELF-TEST, PF, DEVOFFL,
+                                                   UNITOFFL */
+    /* ^ Only set PF(Page format) bit and leave others as 0
+     *   Check SPC-5 rev 7 Table 271 - The meanings of the SELF - TEST CODE
+     *   field, the PF bit, the S ELF T EST bit, and the PARAMETER LIST LENGTH
```

Comment:
is there a particular reason for the spacing between S and T and the rest of the words?